Text/Megaparsec/Class.hs

```diff
@@ -272,6 +273,29 @@ class (Stream s, MonadPlus m) => MonadParsec e s m | m -> e s where
   -- | @'updateParserState' f@ applies the function @f@ to the parser state.
   updateParserState :: (State s e -> State s e) -> m ()
 
+  -- | An escape hatch for defining custom 'MonadParsec' primitives.
+  --
+  -- === __Example__
+  -- @
+  -- breakOn :: Text -> ParsecT e Text m a
+  -- breakOn end = liftConsumer $ \\s ->
+  --   let (pre, post) = T.breakOn end (stateInput s)
+  --    in ( ConsumedOk mempty pre
+  --       , s
+  --           { stateInput = post
+  --           , stateOffset = stateOffset s + T.length pre
+  --           }
+  --       )
+  -- @
+  liftConsumer :: (State s e -> (ParseResult s e a, State s e)) -> m a
+
+-- | A helper containing all the information that 'liftConsumer' can return.
+data ParseResult s e a
+  = ConsumedOk (Set (ErrorItem (Token s))) a
+  | ConsumedErr (ParseError s e)
+  | EmptyOk (Set (ErrorItem (Token s))) a
+  | EmptyErr (ParseError s e)
```

Comment:
This could also be
```hs
(Either (ParseError s e) (Set (ErrorItem (Token s))), a), Bool)
```
But I figured a little duplication + 1 more data type is much nicer than that